### ISSUE TITLE
Channels 2 Multiplexing

### ISSUE THREAD

**Issue by andrewgodwin** [django/channels#825]
Channels 2 needs multiplexing support, as it's not yet implemented.

**Comment by brianmay** [comment]
I am trying to visualize how this should be implemented.

The problem I keep encountering, however, is that the Demultiplexer needs to be the class that terminates the WebSocket connection. As we don't know where to dispatch messages to until we receive them, and every message for a given connection could have a different destination.

So having the Demultiplexer dispatch to a set of Consumer classes seems wrong, because the Consumer classes each assume that they exclusively terminate the websocket connection. They also contain functionality that is not needed, such as reading/writing from the websocket.

So maybe we need a completely independent consumer class (probably sync and async) dedicated to just receiving Demultiplexer JSON messages?

**Comment by andrewgodwin** [comment]
Yes, this is going to need an ASGI application (probably a consumer but might have to be lower-level) that does the de/re-multiplexing itself and then manages a suite of other application instances, one for each stream. It's going to be a lot more complex than the middlewares/routing that just looks at scope.

**Comment by hishnash** [comment]
You think it needs to be this complex?

Why not do as @brianmay suggested and terminate the WebSocket on the `Demultiplexer` and make the child `Consumer`ish classes not be true `Consumer`s but rather be some `DemultiplexedConsumer` the `Demultiplexer` could keep references to them. And they (the `DemultiplexedConsumer` instances) could have weak refs back to the `Demultiplexer` so as to access session info etc. 

with respect to processing messages coming over layers, the `Demultiplexer` could use the same keys as it does for routing WS messages to rout the layer messages. 

```
class APIDemultiplexer(WebsocketDemultiplexer):
    influencer = InfluencerDemultiplexedConsumer
    user = UserDemultiplexedConsumer
```

sending a message `user.action.logout` would call the method `action_logout` on 
 the instance of `UserDemultiplexedConsumer` for this `Demultiplexer`.

 What would the benefit of `de/re-multiplexing itself and then manages a suite of other application instances` bring?

**Comment by andrewgodwin** [comment]
The whole principle of the new ASGI design is turtles all the way down - that is, a demultiplexer should take any valid ASGI application as its demultiplexing endpoints. Otherwise we would need to maintain a whole separate, parallel suite of generic consumers that are these not-quite-consumers, and you wouldn't be able to move things between them, etc.

The pattern is basically:
* When the multiplexer is initialised it also initialises any sub-applications with the same scope
* When its coroutine is started up it also starts up the coroutines of any sub-applications, but passing them new send/receive methods
* When a frame comes in, it demultiplexes it and invokes the appropriate receive awaitable on a child
* When a child invokes their version of send, it multiplexes and sends it on

**Comment by hishnash** [comment]
ok, I see what you mean, still not 100% sure on why it would need to have a separate coroutine for the sub-apps through? I am correct in assuming that at the moment none of the routers etc does this currently?

> When the multiplexer is initialised it also initialises any sub-applications with the same scope

why not initialise sub apps on the first message that matches?

**Comment by andrewgodwin** [comment]
`await_many_dispatch` will be the wrong model - there are other ways of making sure child coroutines close when their parent does. The reason the routers don't need this is that they don't transform incoming and outgoing messages; it's when you do that you need to override both phases (init and coroutine) rather than just the initialisation.

New ASGI was specifically designed to allow this use case, and also to let most common overrides, like routing and auth middleware, not need to do the whole coroutine dance.

**Comment by hishnash** [comment]
So the aim would not be for the sub `Consumer`s to be any subclasses for `Consumer`? since during the base class `__call__` method of `Consumer` it calls `await_many_dispatch`. 

am I correct in assuming `await_many_dispatch` will block that `async` call stack so we would still need a `child` coroutine for each child for the Demultiplexer? after all if they have the same coroutine it will block with the `while True` when the first childs `__call__` is called, so even if the `Demultiplexing`'s `__call__` is changed to not use `await_many_dispatch` it will still be limited to just one child without creating a child `coroutine` for each child?

or is it possible to call the `__call__` method of the children on the same `coroutine` of the parent without blocking the parent? and thus let exceptions etc bubble up without a load of pain.

**Comment by andrewgodwin** [comment]
No, there's no need to "gather" loops, everything is being awaited (including `await_many_dispatch`) so it will all just run. Your previous comment's conclusion, that `await_many_dispatch` needs its own loop, is incorrect.

I'd suggest going and reading up on how asyncio event loops work, especially with nested coroutines and futures, before you attack a problem like this ticket - I don't think having a discussion here about the intricacies of async is going to help (the pattern I outlined earlier should still work just fine).

**Comment by hishnash** [comment]
ok, I believe I have it working as you mentioned above.

 However, if we just try to intercept the upstream `send` messages we then for jsonwebsocket, for example, will need to `decode` them back to json add the `"stream":...` and re-encode to send them up.

https://github.com/hishnash/channels/blob/Re-Implement-Multiplexing-%23825/channels/multiplexing.py#L108-L116 (this is still in a dirty state i will clean up more before doing a PULL Request)

should we do this, in the same way, messages that are sent upstream are handled?

eg.

when we get a downstream (back to the client/web browser etc) message look for a local method that matches the `type` key. 

eg the `AyncJSONWebsocketDemultiplexer` would have a method `websocket_send` that would then get fired when it gets a downstream message from one of its child applications?

**Comment by stunaz** [comment]
any update on this? or is it possible to migrate from channel 1.x.x to channel 2.0.x without this implemented?

**Comment by ruralscenery** [comment]
Before merge multiplexing branch to master, you can try it

install from:
git clone https://github.com/django/channels/archive/multiplexing.zip

in the routing.py:
```
from channels.generic.websocket import WebsocketMultiplexer

EchoMultiplexer = WebsocketMultiplexer({
    "async": EchoConsumerAsync,
    "sync": EchoConsumerSync,
})

application = ProtocolTypeRouter({
    "websocket":
    AuthMiddlewareStack(
        URLRouter([
            path("", EchoMultiplexer),
        ])),
})
```

**Comment by maryokhin** [comment]
@ruralscenery what is this multiplexer implementation? is it for Channels 2.0? Is there an example of `EchoConsumerAsync` child implementation?

**Comment by andrewgodwin** [comment]
This is probably not going to happen unless someone else works on it - I don't have the time any more.

**Comment by hishnash** [comment]
@sevdogI have a freestanding package https://github.com/hishnash/channelsmultiplexer that adds a minimal version of this to channelsv2.

**Comment by sevdog** [comment]
@hishnash I will give a look at your code and maybe try it out.

However I think that this feature should get more support cause it is "essential" to fully upgrade an existing project from channels 1.x to 2.x, because an heavy usage of multiplexing (_multi purpose connection_) is not easy to replace with a set of _single purpose_ socket connections.

**Comment by PapiCZ** [comment]
Any progress? I would really apreceate multiplexing support in channels 2. I don't want to have separate TCP connection for each realtime component in my web app.

**Comment by ronnievdc** [comment]
I did some work to create a multiplex channels consumer: https://gist.github.com/ronnievdc/dd65a1ba1fef86ba73e37faeaaf06de2

Assume you have 2 websockets on one page and want to expose both consumers / channels over one websocket connection
```js
const channel_ws = new Websocket('ws:///url/to/consumer');
const other_channel_ws = new Websocket('ws://url/to/other/consumer');
```

To combine both consumers (and all other consumers you already have defined in **routing.py**)
Add the following line to your routing.py
```python
re_path(r'ws/multiplex/$', AsyncMultiplexWebsocketConsumer),
```

And change your javascript so it uses both consumers over 1 websocket connection, without changing the consumers at all.
```js
const ws = WebSocketMultiplex(new Websocket('ws://ws/multiplex/'));
const channel_ws = ws.channel('/url/to/consumer');
const other_channel_ws = ws.channel('/url/to/other/consumer');
```

**Comment by johnthagen** [comment]
Do any of the recent [`async` enhancements in Django 3.1](https://docs.djangoproject.com/en/stable/releases/3.1/#asynchronous-views-and-middleware-support) or the work towards ASGI v3 (#1480) make official multiplexing support easier to realize?

**Comment by andrewgodwin** [comment]
Multiplexing was never about the underlying protocol or support - the sticking point for me has always been that it needs a lot of client-side code and a good specification that can handle connection drops to make it work well. It can also be totally accomplished as a third-party app, which was always my suggestion if someone wanted to work on it and prove it out before making a pitch to have it rolled in.

**Comment by csdenboer** [comment]
I wrote a third party app that provides a (de)multiplexer for Channels 2: https://github.com/csdenboer/channels-demultiplexer. Is it OK if I create a PR that adds a link to this package to the documentation?

**Comment by huynguyengl99** [comment]
Maybe a bit late here, but as an extension of django-channels, [chanx](https://github.com/huynguyengl99/chanx) has recently added support for multiplexing, under the name 'Topic' feature. Hope it works for anyone who comes here looking for a solution to this."
